c_glib/arrow-glib/file-system.cpp

```diff
@@ -0,0 +1,1092 @@
+
+G_DEFINE_TYPE(GArrowSlowFileSystem,
+              garrow_slow_file_system,
+              GARROW_TYPE_FILE_SYSTEM)
+
+static void
+garrow_slow_file_system_init(GArrowSlowFileSystem *file_system)
+{
+}
+
+static void
+garrow_slow_file_system_class_init(GArrowSlowFileSystemClass *klass)
+{
+}
+
+/**
+ * garrow_slow_file_system_new_average_latency:
+ * @base_file_system: A #GArrowFileSystem as the base file system.
+ * @average_latency: The average value of the latency.
+ *
+ * The latency is normally distributed with a standard deviation of
+ * @average_latency * 0.1.
+ *
+ * The random seed is given by the default random device.
+ *
+ * Returns: (transfer full): A newly created #GArrowSlowFileSystem.
+ *
+ * Since: 1.0.0
+ */
+GArrowSlowFileSystem *
+garrow_slow_file_system_new_average_latency(GArrowFileSystem *base_file_system,
+                                            gdouble average_latency)
+{
+  auto arrow_base_file_system = garrow_file_system_get_raw(base_file_system);
+  auto arrow_slow_file_system =
+    std::make_shared<arrow::fs::SlowFileSystem>(arrow_base_file_system, average_latency);
+  std::shared_ptr<arrow::fs::FileSystem> arrow_file_system = arrow_slow_file_system;
+  auto file_system = garrow_file_system_new_raw(&arrow_file_system,
+                                                GARROW_TYPE_SLOW_FILE_SYSTEM);
+  return GARROW_SLOW_FILE_SYSTEM(file_system);
+}
+
+/**
+ * garrow_slow_file_system_new_average_latency_and_seed:
+ * @base_file_system: A #GArrowFileSystem as the base file system.
+ * @average_latency: The average value of the latency.
+ * @seed: A random seed.
+ *
+ * The latency is normally distributed with a standard deviation of
+ * @average_latency * 0.1.
+ *
+ * Returns: (transfer full): A newly created #GArrowSlowFileSystem.
+ *
+ * Since: 1.0.0
+ */
+GArrowSlowFileSystem *
+garrow_slow_file_system_new_average_latency_and_seed(GArrowFileSystem *base_file_system,
+                                                     gdouble average_latency,
+                                                     gint32 seed)
+{
+  auto arrow_base_file_system = garrow_file_system_get_raw(base_file_system);
+  auto arrow_slow_file_system =
+    std::make_shared<arrow::fs::SlowFileSystem>(arrow_base_file_system, average_latency, seed);
+  std::shared_ptr<arrow::fs::FileSystem> arrow_file_system = arrow_slow_file_system;
+  auto file_system = garrow_file_system_new_raw(&arrow_file_system,
+                                                GARROW_TYPE_SLOW_FILE_SYSTEM);
+  return GARROW_SLOW_FILE_SYSTEM(file_system);
+}
+
+G_END_DECLS
+
+GArrowFileStats *
+garrow_file_stats_new_raw(const arrow::fs::FileStats& arrow_file_stats)
+{
+  auto file_stats = garrow_file_stats_new();
+  GARROW_FILE_STATS_GET_PRIVATE(file_stats)->file_stats = arrow_file_stats;
+  return file_stats;
+}
+
+GArrowFileSystem *
+garrow_file_system_new_raw(std::shared_ptr<arrow::fs::FileSystem> *arrow_file_system,
+                           GType type)
```

Comment:
How about detecting `GType` from `arrow::fs::FileSystem::type_name()` like we did `garrow_array_new_raw()`?
It's useful when we create a binding for a function that returns a generic `arrow::fs::FileSystem` such as `arrow::fs::FileSystemFromUri()`.